datafusion/expr/src/type_coercion/binary.rs

```diff
@@ -25,9 +25,144 @@ use arrow::datatypes::{
 use datafusion_common::DataFusionError;
 use datafusion_common::Result;
 
-use crate::type_coercion::{is_datetime, is_decimal, is_interval, is_numeric};
+use crate::type_coercion::is_numeric;
 use crate::Operator;
 
+/// The type signature of an instantiation of binary expression
+struct Signature {
+    /// The type to coerce the left argument to
+    lhs: DataType,
+    /// The type to coerce the right argument to
+    rhs: DataType,
+    /// The return type of the expression
+    ret: DataType,
+}
+
+impl Signature {
+    /// A signature where the inputs are coerced to the same type as the output
+    fn coerced(t: DataType) -> Self {
```

Comment:
maybe this could be called "uniform" to better describe it -- coerce I think is the general term for automatic casting.